Adding HMAC support to http_endpoint input

P1llus opened the pull request:
## What does this PR do?

To support certain Webhooks that uses HMAC signatures (like Github) this change introduces the possibility to set a HMAC token, a HMAC header and optionally a HMAC Prefix (if the header includes a prefix before the token).

## Why is it important?

Allows the http_endpoint input to support more security features

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates #19534

P1llus commented:
jenkins test this please

marc-gr commented on x-pack/filebeat/input/http_endpoint/validate.go:
```diff
@@ -52,3 +73,33 @@ func (v *apiValidator) ValidateHeader(r *http.Request) (int, error) {
 }
+
+func (v *apiValidator) ValidateHmac(r *http.Request) (int, error) {
+	if v.hmacHeader != "" && v.hmacToken == "" {
```
Just to nit pick, since `hmacHeader` seems to be mandatory I would early return if empty, then check for `hmacToken` being empty, and then doing the rest ouside of any if, IMO the logic would look a bit more clear, but feel free to ignore the comment.

P1llus replied:
Hmm they are both mandatory, but you are correct, the code also only checks if header is set but not token, and not the other way around. They both need to be configured because the HMAC signature needs to be included somewhere in the request. So if token != null, we still have to specify where the token is, and the other way around.

Have you ever seen usecases in which HMAC would be included somewhere else? Like a JSON field?

marc-gr commented on x-pack/filebeat/input/http_endpoint/validate.go:
```diff
@@ -5,15 +5,22 @@
 )
 
 type validator interface {
+	Validate(*http.Request) (int, error)
```
To me it is a bit confusing to have an interface which defines methods that are meant to be called from eachother instead of by the user, since it is not evident when (or if) `validateHmac` and `validateHeaders` are called. Maybe changing validator to just have the `validate` method instead, and then either doing the validations inside without exposing `validateHmac` and `validateHeaders` through the interface, or having several validators, one for each validation and executing them in series would be a bit better. WDYT?

P1llus replied:
Yeah I think that is a good assumption. At first I wanted to modify the withValidator func in handler.go to include just both validateHmac and validateHeaders, but if we ever want to add more validation steps I rather wanted to keep that part small, adding a generic validate method there instead. I think I can move the rest outside the interface for sure, still getting used to how golang uses them :)

urso replied:
Header and body should be validated separately/differently. Processing the body might require some protecting from large contents as well (e.g. when sending an infinite stream using chunks).

I'd prefer to keep the method `ValidateHeader`, but change the interface to `ValidateHeader(http.Header)`, WDYT?

marc-gr commented on x-pack/filebeat/input/http_endpoint/validate.go:
```diff
@@ -52,3 +73,33 @@ func (v *apiValidator) ValidateHeader(r *http.Request) (int, error) {
+			return http.StatusInternalServerError, fmt.Errorf("The HMAC signature in the configured request header is empty")
+		}
+		s := hmac.New(sha1.New, []byte(v.hmacToken))
+		b, err := ioutil.ReadAll(r.Body)
```
Not sure if reading this here (and not closing it) might cause issues to following attempts of reading it. I think you might need to read all, close the body, and create a new ReadCloser from the read bytes IIRC.

```
b, _ := ioutil.ReadAll(r.Body)
...
r.Body.Close()  //  must close
r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
```

P1llus commented:
After some pointers and discussions with @marc-gr I added a more simplified way to extend and split up validators, much thanks to Marc for the pointers and guidance! :)

urso commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -82,16 +82,25 @@ func (e *httpEndpoint) Test(_ v2.TestContext) error {
+	headerValidator := &headerValidator{
+		method:      http.MethodPost,
+		contentType: e.config.ContentType,
+	}
```
The validators need to check on every request if they are configured or not. Worse, they even check for missconfiguration at runtime (users can detect beats miss configuration in HTTP client at runtime only). I'd propose to separate configuration time from runtime, and build an array of configured/enabled header validators (e.g. `validatorList` type shown in another comment). Faulty configurations must be detected when creating the endpoint object, not maybe later after running it.

urso commented on x-pack/filebeat/input/http_endpoint/validate.go:
```diff
@@ -5,43 +5,59 @@
+type headerValidator struct {
+	method      string
+	contentType string
+}
```
all validators are essentially header validators. Consider to rename this one into method validator or  split this one into method and content-type validators.

urso commented on x-pack/filebeat/input/http_endpoint/validate.go:
```diff
@@ -52,3 +68,35 @@ func (v *apiValidator) ValidateHeader(r *http.Request) (int, error) {
+	}
+
+	return 0, nil
+}
```
hm, this is doing more than just validating the contents. This looks like an intermediate data processing level. Maybe this should be some middleware layer that wraps the body. The custom reader would stream all writes from body to the hmac 'validator'. If we hit EOF, the reader does the final check and fails if there was a failure. In case the wrapped handler did not consume the body, the middleware handler would read until EOF (or error) in order to enforce the HMAC check.

Using layers we would create a processing chain like: `bodyLimit -> hmac -> header validation -> body validation`. The `bodyLimit` would wrap the body and fail in case the body (or contents-length) is bigger then the configured limits.

Are there packages/projects providing authentication support as middleware, we can use?
